### PR TITLE
Create tm webhooks during ord aggregation

### PR DESCRIPTION
**Pull request by alextargov**
**Description**

Changes proposed in this pull request:
- ...

**Related issue(s)**

- ...

**Pull Request status**

- [x] Implementation
- [x] Unit tests
- [ ] Integration tests
- [x] `chart/compass/values.yaml` is updated 
- [x] Mocks are regenerated, using the automated script

### CODE REVIEW

File: components/director/internal/domain/webhook/service.go [hunk 1]
```diff
@@ -188,6 +202,75 @@ func (s *service) Delete(ctx context.Context, id string, objectType model.Webhoo
...
+				HeaderTemplate: &headerTemplate,
+				OutputTemplate: tenantMappingWebhook.OutputTemplate,
+			}
+			webhooks = append(webhooks, wh)
+		}
+	}
+	return webhooks, nil
+}
+
+func (s *service) getTenantMappingWebhooks(mode, version string) ([]graphql.WebhookInput, error) {
```
Comment (dragobt) [inline components/director/internal/domain/webhook/service.go:247]: add godoc

Comment (dragobt) [inline components/director/internal/domain/webhook/service.go:247]: add godoc

File: components/director/internal/open_resource_discovery/service.go [hunk 2]
```diff
@@ -1617,3 +1799,35 @@ func addFieldToLogger(ctx context.Context, fieldName, fieldValue string) context
 	logger = logger.WithField(fieldName, fieldValue)
 	return log.ContextWithLogger(ctx, logger)
 }
+
+func createWebhookInput(credentialExchangeStrategyJSON gjson.Result, tenantMappingData CredentialExchangeStrategyTenantMapping) *graphql.WebhookInput {
+	inputMode := graphql.WebhookMode(tenantMappingData.Mode)
+	return &graphql.WebhookInput{
+		URL: str.Ptr(credentialExchangeStrategyJSON.Get(callbackURLProperty).String()),
+		Auth: &graphql.AuthInput{
+			AccessStrategy: str.Ptr(string(accessstrategy.CMPmTLSAccessStrategy)),
```
Comment (dragobt) [inline components/director/internal/open_resource_discovery/service.go:1808]: Redundant string conversion

File: components/director/internal/open_resource_discovery/service.go [hunk 3]
```diff
@@ -768,6 +823,133 @@ func (s *Service) resyncBundleInTx(ctx context.Context, appID string, bundlesFro
...
+	if err != nil {
+		return err
+	}
+
+	isEqual, err := isWebhookDataEqual(tenantMappingRelatedWebhooksFromDB, enrichedWhModels)
+	if err != nil {
+		return err
+	}
+
+	if isEqual {
```
Comment (dragobt) [inline components/director/internal/open_resource_discovery/service.go:869]: can we write a unit test when this is true

File: components/director/internal/open_resource_discovery/service.go [hunk 4]
```diff
@@ -768,6 +823,133 @@ func (s *Service) resyncBundleInTx(ctx context.Context, appID string, bundlesFro
...
+}
+
+func (s *Service) deleteWebhooks(ctx context.Context, webhooks []*model.Webhook, appID string) error {
+	for _, webhook := range webhooks {
+		log.C(ctx).Infof("Deleting webhook with ID %s for application %s", webhook.ID, appID)
+		if err := s.webhookSvc.Delete(ctx, webhook.ID, webhook.ObjectType); err != nil {
+			log.C(ctx).Errorf("error while deleting webhook with ID %s", webhook.ID)
+			return errors.Wrapf(err, "while deleting webhook with ID %s", webhook.ID)
+		}
+	}
```
Comment (dragobt) [inline components/director/internal/open_resource_discovery/service.go:894]: this logic is not tested

File: components/director/internal/open_resource_discovery/service.go [hunk 5]
```diff
@@ -768,6 +823,133 @@ func (s *Service) resyncBundleInTx(ctx context.Context, appID string, bundlesFro
...
+
+	return nil
+}
+
+func (s *Service) getTenantMappingData(credentialExchangeStrategyJSON gjson.Result, appID string) (CredentialExchangeStrategyTenantMapping, error) {
+	tenantMappingType := credentialExchangeStrategyJSON.Get(customTypeProperty).String()
+	tenantMappingData, ok := s.config.credentialExchangeStrategyTenantMappings[tenantMappingType]
+	if !ok {
+		return CredentialExchangeStrategyTenantMapping{}, errors.Errorf("Credential Exchange Strategy has invalid %s value: %s for application with ID %s", customTypeProperty, tenantMappingType, appID)
```
Comment (dragobt) [inline components/director/internal/open_resource_discovery/service.go:915]: unit test here

File: components/director/internal/open_resource_discovery/service.go [hunk 6]
```diff
@@ -768,6 +823,133 @@ func (s *Service) resyncBundleInTx(ctx context.Context, appID string, bundlesFro
...
+
+		for _, webhookFromDB := range webhooksFromDB {
+			if webhookFromDB.Type == convertedIn.Type {
+				webhookModel.ID = webhookFromDB.ID
+				webhookModel.ObjectType = webhookFromDB.ObjectType
+				webhookModel.ObjectID = webhookFromDB.ObjectID
+				webhookModel.CreatedAt = webhookFromDB.CreatedAt
+
+				tenantMappingRelatedWebhooksFromDB = append(tenantMappingRelatedWebhooksFromDB, webhookFromDB)
+				break
```
Comment (dragobt) [inline components/director/internal/open_resource_discovery/service.go:943]: this logic is also not tested

File: components/director/internal/domain/webhook/service.go (hunk 1 above)
Comment (alextargov) [reply components/director/internal/domain/webhook/service.go:247]: We do not place godoc to unexported functions

File: components/director/internal/open_resource_discovery/service.go [hunk 7]
```diff
@@ -768,6 +823,133 @@ func (s *Service) resyncBundleInTx(ctx context.Context, appID string, bundlesFro
 	return tx.Commit()
 }
 
+func (s *Service) resyncTenantMappingWebhooksInTx(ctx context.Context, credentialExchangeStrategyJSON gjson.Result, appID string) error {
+	if !credentialExchangeStrategyJSON.IsObject() {
+		log.C(ctx).Debugf("There are no tenant mappings to resync")
```
[the remark this answers is not in the document]
Comment (alextargov) [reply components/director/internal/open_resource_discovery/service.go:828]: I think we have discussed that. If credentialExchangeStrategyJSON is empty, then there are no tenant mappings to create.

File: components/director/internal/open_resource_discovery/service.go (hunk 2 above)
Comment (alextargov) [reply components/director/internal/open_resource_discovery/service.go:1808]: Not redundant. IDE says it is, but at runtime, an error is thrown. I also thought that it is redundant but apparently, it is not. :( 
image

**Review by PetarTodorovv — APPROVED** [review]

**Review by PetarTodorovv — APPROVED** [review]

**Comment by alextargov** [comment]
reset-vm

**Pull request merged**